### PR TITLE
feat: Sonarr Import List mode

### PR DESCRIPTION
**Pull request by Rojikku**
# Import List Mode
The long awaited Import List Mode has arrived!
Sleep is for the week. Which week? Next one, obviously.

- [X] Import List Mode fully working in Sonarr
- [X] Full nginx server because writing webhosting code is hard and unnecessary

When combined with RESPECTFUL_ADDING, this should, and mostly does minus some unidentified bugs, create a list of all anime we want to add, and then remove the entries that are already added on every refresh.

image

Note: Your URL may vary, that is my legitimate URL though in my use case.

I know what you're thinking. "But Rojikku, I want to import on every refresh, not every six hours!"
Fortunately, do I have just the thing for you, my impatient friends. It uses the API to force a refresh on every load! Therefore, auto-matching the interval from AniPlanrr. Additionally, you can set all sorts of nice settings straight in the import list settings, and not have to mess with them in code at all!

Radarr uses a different format so I'm neglecting it for the time being. It's not necessarily harder, it's just late at night and I'm the big sleepy.

P.S. You may have noticed I used a json= in my post request. If I did not do this, the API call did not work. I might recommend investigating any other calls you use that don't do this.

### CODE REVIEW

**Comment by noggl** [comment]
It seems like this requires sonarr to be version 4 or later, is that correct? When I try to add this import list at /settings/importlists, my options are Trakt, Plex, and Sonarr.

Just want to confirm before I backup my system and test this new feature this week before merging

**Comment by Rojikku** [comment]
That...might well be the case. I'll double check my version number later
tonight.
I'm running the latest, and it might well be 4. Custom List is on the
bottom for me, so definitely not one of those options.

**Comment by Rojikku** [comment]
Update: I am running v4.0.0.539, so you are correct.

**Comment by noggl** [comment]
thanks for confirming. I'm running V4 now and will test this week/weekend. If I find any bugs I'll report them on your fork.

Since this is only compatible with V4 though, I'll keep the new feature and the current repo on different branches and make the github action create separate docker images. Once popular container maintainers (like linuxserver.io) switch to V4 for their default version I can merge it into one branch.

**Comment by Rojikku** [comment]
You're welcome to organize this as you please, however, it's entirely
unnecessary to have a separate container/branch for this feature.
Assuming I implemented it correctly- and feel free to report a bug if I
didn't- the previous functionality is both intact and the default.

This is simply an optional additional feature users can use, which IMO
works a lot better thanks to how much Sonarr supports it.
Would be less admin burden on you to just keep it in one and have the flag
required to turn on this mode. If v4 becomes a default later we can phase
out the old mode as desired, or change the default.

**Comment by reitenth** [comment]
is this coming to stable anytime soon? pretty great feature.

**Comment by noggl** [comment]
I'll do some more testing tonight and should be able to merge or request changes for bug fixes. So far everything was working pretty well.

**Comment by Rojikku** [comment]
> is this coming to stable anytime soon? pretty great feature.

Aww, thanks! Hopefully! Just waiting on review.

A few notes for review, while I'm thinking about it, since I've been running it this entire time.

1. Radarr is not implemented. I googled their formatting for like 20 minutes and I didn't see it. Might be the same, I recall seeing it once and it was a bit different.
~~2. It only adds one Anime at a time, and it seems to wait until it's downloaded before adding the next one. I'm assuming this is intentional from Sonarr. That, or it's only reading the first entry on the website, which will change as series are added. But I my formatting is picture perfect for their example, so unless they changed it without note... Should be intentional. Easy fix either way.~~ Edit: That was my bug, fixed. Python is hard.
3. I think I mentioned some possible issues with matching before. No wrong series get added by any means, but some series I already have added do get attempted to add. I haven't spent much effort investigating because I don't believe it causes issues and I don't believe it's hindered me. ~~But if the above is a bug with only reading the first entry, it could cause adding to hang on attempting to add an existing entry, in theory.~~ EDIT: Irrelevant, it totally did become an issue though.

I could probably investigate and see if I see that issue on my instance after work. Either way, code works well enough for my instance. Fine tuning can be done later as needed. It's pretty simple to modify once it's implemented. Same for copying to Radarr, mostly copy paste and formatting if we find the format for it again and want it.
No noticeable issues as a whole, though, beyond noticing that delay with adding.

**Comment by Rojikku** [comment]
> I'll do some more testing tonight and should be able to merge or request changes for bug fixes. So far everything was working pretty well.

Tee-hee? I found a bug. It was mine.
Apparently I had to declare my little json formatting variable within the loop otherwise python magic happened and it updated every instance of that same variable that was already appended to the list... It now happily adds five at once with the latest fix I added.

Side note: It appears TVDBID 429310 is specifically the trouble child for my mapping list.

**Comment by noggl** [comment]
I had a weird bug where I got an email that the series had been deleted seconds before it was added. I haven't been able to figure out where that came from yet. It doesn't seem to be a problem, but it was pretty weird.

Screen Shot 2023-06-27 at 6 27 07 AM

**Comment by Rojikku** [comment]
> I had a weird bug where I got an email that the series had been deleted seconds before it was added. I haven't been able to figure out where that came from yet. It doesn't seem to be a problem, but it was pretty weird.
> 
> Screen Shot 2023-06-27 at 6 27 07 AM
> 


Hmm. That's definitely weird.
That would have to be a bug on Sonarr's end though. My code only hosts a webpage and (tries to) trigger a sync of import lists. Sonarr sometimes ignores the command though, when it's busy. I can't see how either action would cause that. 

I don't do email notifications, so can't replicate... I will say none of my series with content have had any issues.

**Comment by noggl** [comment]
I've done import lists in sonarr before and they don't cause this, so it does seem to be something specifically with this list.

I also have had no issues with the content itself, it's just this strange email.

If you're able to turn on email notifications to confirm that would be helpful, otherwise I'll continue to test it on my own.

**Comment by Rojikku** [comment]
> I've done import lists in sonarr before and they don't cause this, so it does seem to be something specifically with this list.

Hmm. Issue with the custom list format, which is new, or perhaps one of their new libraries in v4?  

> I also have had no issues with the content itself, it's just this strange email.
> 
> If you're able to turn on email notifications to confirm that would be helpful, otherwise I'll continue to test it on my own.

Ah, can't. I have all sorts of security setup on my email that won't allow me to send stuff from other sources. I have to use discord for notifications on everything.

Could potentially test with those, but any deletions without explicit api commands to do so would still be a sonarr bug.
Can test this evening, but I don't know what you'd gain from the information. 

All I do is feed it a list of IDs in a text file, and tell it to look at it. If an ID is there that's already in the system, it's unintended on my part because I'm orderly, but it should be something sonarr is used to because all other lists should work like that.

In other words, there's no situation where an ID being or not being on the list should cause anticipated issues.

Though there's always the exclusion list if you want something to never be added again.

**Comment by noggl** [comment]
I am still testing this, This morning I was getting 7 occurrences of the same show when I went to the /sonarr endpoint using the code from yesterday (after 1 round of successfully adding). I just refreshed it and now I'm getting 7 different shows, correctly. There is not functionality for specific seasons using this import list mode, right? It's based solely on the monitor setting for the custom list in sonarr's UI?

I'm not sure that the immediate list refresh is working properly. I've got 7 shows (5 that are already in sonarr) visible at the /sonarr endpoint, but sonarr isn't grabbing them.
```
==> aniplanrr.log <==
Running aniplanrr.py
Logging is disabled
Loaded 4919 items to map
3-gatsu no Lion is mapped to 3-gatsu no Lion season 1
Ima, Soko ni Iru Boku is mapped to Ima, Soko ni Iru Boku season 1
Jujutsu Kaisen 2nd Season is mapped to Jujutsu Kaisen season 2
Kidou Senshi Gundam: Suisei no Majo is mapped to Kidou Senshi Gundam: Suisei no Majo season 1
Tengoku Daimakyou is mapped to Tengoku Daimakyou season 1
Buddy Daddies is mapped to Buddy Daddies season 1
Kaminaki Sekai no Kamisama Katsudou is mapped to Kaminaki Sekai no Kamisama Katsudou season 1
Found 7 new shows to add to Sonarr (I uncommented this specific printline to check something earlier)
Wrote sonarr
Sent command to Sonarr to import new shows
Movie PERFECT BLUE already in Radarr
Movie Seishun Buta Yarou wa Yumemiru Shoujo no Yume wo Minai already in Radarr
Searching for Komada Jouryuusho e Youkoso by title and year
ID not received from radarr for Komada Jouryuusho e Youkoso
Sync Completed
Sleeping for 3600 seconds

==> sonarr.log <==
[Info] ImportListSyncService: Starting Import List Sync 
[Info] ImportListSyncService: Processing 0 list items 
[Info] ImportListSyncService: Import List Sync Completed. Items found: 0, Series added: 0 
```

I am also still looking at what's causing that email, since I don't actually have "Series Deleted" turned on as a notification type. I'm not so concerned about that one though, it does seem like some weird sonarr bug.

**Comment by Rojikku** [comment]
The 7 occurrences of the same show was a bug I fixed last night on this PR.
Comment above on it. Glad it's working now though.

The specific seasons with this import list mode is not really a thing, you
are correct, purely based on sonarr's UI. Would be a feature request for
them. Personally I'd rather easily set my own quality profile than set
seasons, so this is to my advantage. But keeping both modes on the program
to provide that availability makes sense.

The immediate list refresh is not working properly. To explain:
1. I trimmed down the code to ONLY have the refresh when I made the PR. I
got it working and verified it was making Sonarr happy. Json= instead of
data= was a key part of it registering.
2. I put the code back, and it seemed to work.
3. I noticed it wasn't working, similar to how you did, yesterday.
4. I started debugging, and I started where I started last time- I put in a
curl of the raw command from a working example. It...also didn't work. So
the issue is likely with the Sonarr API and probably needs a bug report. I
had thought it was because I had just added 5 series, and that was the
issue (queue too long)- but seems not if you've replicated it.
5. You can still manually do a refresh by opening the import list in
sonarr, and hitting save.

There's not anything I can do about it since it seems to be an API bug.
It's sending the command properly, far as I can tell, and I can't find a
better working example... If you do, though, I'm happy to adjust for that
and improve the code.
I swear it was working a few versions ago, recent changes could have broken
it. It also could be because I only checked if sonarr got the command to
refresh, not if it actually did it right. But the command producing a
different result from the UI triggering the same command should be a bug
report, still.
I tried finding other bug reports, one said charset=utf8 was necessary, but
even adding that to my curl line didn't produce the desired result, so... I
don't know.

The fact you don't have that notification type turned on is interesting,
even more likely a bug.

**Comment by noggl** [comment]
The "Series Deleted" bug was [indeed a bug with sonarr](https://github.com/Sonarr/Sonarr/commit/5c48049702269276a43a54531b3377a6cff22d0c) that is now fixed. I've had no issues with your feature since [2480acb](https://github.com/noggl/AniPlanrr/pull/11/commits/2480acb64881824818adbe6813eb3e8f10c66554), so I'll merge this.

I'll probably reformat the readme this evening to introduce the two modes more clearly.

**Comment by reitenth** [comment]
wat should the list url be? thanks.

**Comment by Rojikku** [comment]
> wat should the list url be? thanks.

It depends on your environment and how you set it up. As a whole, should be the same as my example, but with the IP of your sonarr box.
Though you can do docker magic to change the port easily.

**Comment by reitenth** [comment]
image

**Comment by Rojikku** [comment]
> image
> 

You can probably just web browse to the appropriate URL and know it when you see it.

If you go to that one, you should see the main page, which should explicitly state you probably want `/sonarr` as in my example in the first post.

I'll assume you changed the port intentionally and know what you're doing.

Also will assume you have proper DNS setup and can curl the URL you're trying to use from your sonarr instance if you exec into it.

Theoretically those are all the troubleshooting steps you should ever need. Good luck!

Edit: Looking closely, I see connection refused... So you probably want to find the right IP or port.

My DNS name for aniplanrr is aniplanrr, so I can do that nonsense. But might not work in all environments!

**Comment by reitenth** [comment]
how do i find the right ip of the container if it doesnt have an ip? 
 
image

**Comment by reitenth** [comment]
added the sonarrimporter and desired port according to readme, didnt have those before

**Comment by Rojikku** [comment]
>  how do i find the right ip of the container if it doesnt have an ip? 
>  
> image
> 

... I don't know why you're asking me. You got aniplanrr to contact sonarr. This is the same thing but the other way around.

Just do what you did to get its IP.
